File: verify_block_matching.c
#include <stdio.h>
#include "block_matching.h"

// Test function to compare C and Python implementations
void test_block_matching() {
    // Create test images
    int height = 64;
    int width = 64;
    Image* ref_img = create_image(height, width);
    Image* target_img = create_image(height, width);

    // Fill with test pattern
    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            // Reference image: diagonal gradient
            ref_img->data[y * width + x] = (float)(x + y) / (width + height);
            
            // Target image: shifted diagonal gradient
            int shift_x = 2;
            int shift_y = 3;
            int tx = (x + shift_x) % width;
            int ty = (y + shift_y) % height;
            target_img->data[y * width + x] = (float)(tx + ty) / (width + height);
        }
    }

    // Setup block matching parameters
    BlockMatchingParams params = {
        .num_levels = 3,
        .factors = (int[]){1, 2, 4},
        .tile_sizes = (int[]){8, 16, 32},
        .search_radii = (int[]){4, 4, 4},
        .use_l1_dist = (bool[]){true, true, true},
        .pad_top = 0,
        .pad_bottom = 0,
        .pad_left = 0,
        .pad_right = 0,
        .debug_mode = true
    };

    // Initialize pyramid
    ImagePyramid* ref_pyramid = init_block_matching(ref_img, &params);
    
    // Perform block matching
    AlignmentMap* alignments = align_image_block_matching(target_img, ref_pyramid, &params);

    // Print results
    printf("Alignment Results:\n");
    for (int y = 0; y < alignments->height; y++) {
        for (int x = 0; x < alignments->width; x++) {
            printf("Patch (%d,%d): dx=%.1f dy=%.1f\n", 
                   y, x,
                   alignments->data[y * alignments->width + x].x,
                   alignments->data[y * alignments->width + x].y);
        }
    }

    // Cleanup
    free_image(ref_img);
    free_image(target_img);
    free_image_pyramid(ref_pyramid);
    free_alignment_map(alignments);
}

int main() {
    test_block_matching();
    return 0;
} 

File: verify_block_matching.py
import numpy as np
from block_matching import init_block_matching, align_image_block_matching

def test_block_matching():
    # Create test images
    height = 64
    width = 64
    
    # Create reference image: diagonal gradient
    ref_img = np.zeros((height, width), dtype=np.float32)
    for y in range(height):
        for x in range(width):
            ref_img[y,x] = float(x + y) / (width + height)

    # Create target image: shifted diagonal gradient
    shift_x = 2
    shift_y = 3
    target_img = np.roll(np.roll(ref_img, shift_y, axis=0), shift_x, axis=1)

    # Setup parameters
    options = {
        'verbose': 2
    }
    
    params = {
        'tuning': {
            'factors': [1, 2, 4],
            'tileSizes': [8, 16, 32],
            'distances': ['L1', 'L1', 'L1'],
            'searchRadia': [4, 4, 4]
        }
    }

    # Initialize pyramid
    ref_pyramid = init_block_matching(ref_img, options, params)
    
    # Perform block matching
    alignments = align_image_block_matching(target_img, ref_pyramid, options, params)

    # Print results
    print("Alignment Results:")
    for y in range(alignments.shape[0]):
        for x in range(alignments.shape[1]):
            print(f"Patch ({y},{x}): dx={alignments[y,x,0]:.1f} dy={alignments[y,x,1]:.1f}")

if __name__ == "__main__":
    test_block_matching() 